cpp/src/rolling/rolling.cu

```diff
@@ -623,5 +625,216 @@ std::unique_ptr<column> rolling_window(column_view const& input,
+    thrust::make_transform_iterator(thrust::make_counting_iterator<size_type>(0), following_calculator),
+    min_periods, aggr, mr
+  );
+}
+
+namespace
+{
+  bool is_supported_range_frame_unit(cudf::data_type const& data_type) {
+    auto id = data_type.id();
+    return id == cudf::TIMESTAMP_DAYS
+        || id == cudf::TIMESTAMP_SECONDS
+        || id == cudf::TIMESTAMP_MILLISECONDS
+        || id == cudf::TIMESTAMP_MICROSECONDS
+        || id == cudf::TIMESTAMP_NANOSECONDS;
+  }
+
+  size_t multiplication_factor(cudf::data_type const& data_type) {
+    // Assume timestamps.
+    switch(data_type.id()) {
+      case cudf::TIMESTAMP_DAYS         : return 1L;
+      case cudf::TIMESTAMP_SECONDS      : return 24L*60*60;
+      case cudf::TIMESTAMP_MILLISECONDS : return 24L*60*60*1000;
+      case cudf::TIMESTAMP_MICROSECONDS : return 24L*60*60*1000*1000;
+      default  : 
+        CUDF_EXPECTS(data_type.id() == cudf::TIMESTAMP_NANOSECONDS, 
+                    "Unexpected data-type for timestamp-based rolling window operation!");
+        return 24L*60*60*1000*1000*1000;
+    }
+  }
+
+  template <typename TimestampImpl_t>
+  std::unique_ptr<column> grouped_time_range_rolling_window_impl( column_view const& input,
+                                                      column_view const& timestamp_column,
+                                                      rmm::device_vector<cudf::size_type> const& group_offsets,
+                                                      rmm::device_vector<cudf::size_type> const& group_labels,
+                                                      size_type preceding_window_in_days, // TODO: Consider taking offset-type as type_id. Assumes days for now.
+                                                      size_type following_window_in_days,
+                                                      size_type min_periods,
+                                                      std::unique_ptr<aggregation> const& aggr,
+                                                      rmm::mr::device_memory_resource* mr) {
+
+    TimestampImpl_t mult_factor {static_cast<TimestampImpl_t>(multiplication_factor(timestamp_column.type()))};
+  
+    auto preceding_calculator = 
+      [
+        d_group_offsets = group_offsets.data().get(),
+        d_group_labels  = group_labels.data().get(),
+        d_timestamps    = timestamp_column.data<TimestampImpl_t>(),
+        preceding_window_in_days,
+        mult_factor
+      ] __device__ (size_type idx) {
+        auto group_label = d_group_labels[idx];
+        auto group_start = d_group_offsets[group_label];
+        auto lower_bound = d_timestamps[idx] - preceding_window_in_days*mult_factor;
+
+        auto preceding_i{idx}; // Start at current row.
+        // If preceding value is part of current group, and also exceeds lower_bound, extend window backwards.
+        while ((preceding_i-1) >= group_start && d_timestamps[preceding_i-1] >= lower_bound) {
```

Comment:
I think this could be a binary search. In cases where the search is large, the linear search you have now will be quite expensive. Either one of these two approaches would work.

1) Do a sequential `thrust::lower_bound` (thrust::seq exec policy) for a single value (your `lower_bound`) using reverse_iterator from `idx` to `group_start`.
2) Lift the search out of the rolling window call using a `thrust::lower_bound` call from the host (with a device execution policy) with a vector of lower bounds to search for (d_timestamps with a transform_iterator to offset each timestamp by `preceding_window_in_days*mult_factor`). This would give you a vector of window starts. The downside of this one is you have to materialize that vector -- takes memory and bandwidth. So I would start with option 1.

Reply:
>  a sequential thrust::lower_bound (thrust::seq exec policy) for a single value (your lower_bound) using reverse_iterator from idx to group_start.

Oddly enough, this was where I started. While the unit-tests did actually run, I ran into CUDA launch errors. (I'll have to dig through my notes for the exact stack trace.) I can't recall if I bugged @devavret about what I was running into. Switching to the current implementation seems to have eliminated that problem. 

I'll try this again, and post here.

Reply:
You may have had an exec policy other than thrust::seq, which would mean you are trying to do a dependent kernel launch, which requires compilation with `-rdc`, which we don't do in libcudf.

Reply:
Just tested this out. I can confirm that this works well at a reasonably large test-scale. Thank you, @harrism!